services/apps/projects/src/container/container.service.ts

```diff
@@ -70,19 +105,68 @@ export class ContainerService {
     return existing;
   }
 
-  private toContainer(id: string, projectId: string) {
+  private toContainer(id: string, projectId: string, token: string): ContainerDto {
     return {
       id,
       projectId,
+      token,
       url: this.containerUrl(id),
     };
   }
 
   private containerUrl(id: string): string {
-    return `${environment.docker.proxyHost}/containers/${id.substring(0, 12)}`;
+    // define workspace through folder query parameter /?folder=...
+    return `${environment.docker.proxyHost}/containers/${id.substring(0, 12)}/?folder=${this.codeWorkspace}`;
   }
 
   private idBin(projectId: string) {
     return projectId.slice(-2); // last 2 hex chars
   }
+
+  private checkAllHeartbeats() {
+    const bindPrefix = path.resolve(environment.docker.bindPrefix);
+
+    this.docker.listContainers({
+      filters: {
+        label: [`org.fulib.project`],
+        status: ['created', 'running'],
+      },
+    },
+       (err, containers) => {
+      if (err) {
+        console.log('something went wrong');
+      } else {
+        if(containers?.length) {
+          // loop over all containers, get ProjectId, check heartbeat and remove if needed
+          for(let i = 0; i < containers.length; i++) {
+            let container = containers[i];
+            let projectId = container['Labels']['org.fulib.project'];
+            let p = path.resolve(`${bindPrefix}/heartbeats/${this.idBin(projectId)}/${projectId}/heartbeat`);
+            let check = this.checkHeartbeatMTime(p, projectId);
+
+            if(check == true) {
```

Comment:
I find `check` a little unintuitive -- maybe call the method `isHeartbeatExpired`. Also you can inline the variable and remove the `== true`.